lib/services/android-project-properties-manager.ts

```diff
@@ -0,0 +1,116 @@
+///<reference path="../.d.ts"/>
+"use strict";
+
+import path = require("path");
+
+export class AndroidProjectPropertiesManager implements IAndroidProjectPropertiesManager {
+	private static LIBRARY_REFERENCE_KEY_PREFIX = "android.library.reference.";
+	
+	private _editor: IPropertiesParserEditor = null;
+	private filePath: string = null;
+	private projectReferences: ILibRef[];
+	private dirty = false;
+	
+	constructor(private $propertiesParser: IPropertiesParser,
+		private $fs: IFileSystem,
+		directoryPath: string) {
+			this.filePath = path.join(directoryPath, "project.properties");
+	}
+	
+	public addToPropertyList(key: string, value: string): IFuture<void> {
+		return (() => {
+			let editor = this.createEditor().wait();
+			let i = 1;
+    		while (editor.get(this.buildKeyName(key, i))) {
+        		i++;
+			}
+			
+			editor.set(this.buildKeyName(key, i), value);
+			this.$propertiesParser.saveEditor().wait();
+			this.dirty = true;
+		}).future<void>()();
+	}
+	
+	public removeFromPropertyList(key: string, value: string): IFuture<void> {
+		return (() => {
+			let editor = this.createEditor().wait();
+			let valueLowerCase = value.toLowerCase();
+			let i = 1;
+		    let currentValue: any;
+		    while (currentValue = editor.get(this.buildKeyName(key, i))) {
+		        if (currentValue.toLowerCase() === valueLowerCase) {
+		            while (currentValue = editor.get(this.buildKeyName(key, i+1))) {
+		                editor.set(this.buildKeyName(key, i), currentValue); 
+		                i++;
+		            }
+		            editor.set(this.buildKeyName(key, i));
+		            break;
+		        }
+		        i++;
+		    }
+			this.$propertiesParser.saveEditor().wait();
+			this.dirty = true;
+		}).future<void>()();		
+	}
+	
+	public getProjectReferences(): IFuture<ILibRef[]> {
+		return (() => {
+			if(!this.projectReferences || this.dirty) {
+				let allProjectProperties = this.getAllProjectProperties().wait();
+				this.projectReferences = _(_.keys(allProjectProperties))
```

Comment:
```_(_.keys(allProjectProperties)) ```
could be
```
_(allProjectProperties)
.keys()
.filter....
```

But as you are using the key and value later, consider the following code:
```TypeScript
let libraryReferences: ILibRef[] = [];
_.each(allProjectProperties, (value: any, key: string) => {
     if(_.startsWith(key, "android.library.reference.")) {
            libraryReferences.push(this.createLibraryReference(key, value);
     }
}
```
I'm not sure which one is better, so consider this comment just as suggestion :smile_cat: